### DISCUSSION TITLE
breakout hour prices

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [JaccoR/hass-entso-e#21] [Ideas]
HI,

been running you fine integration for only 2 days now, and already want more ;-)

trying to create graphs like:

<img width="369" alt="Schermafbeelding 2022-10-13 om 09 22 44" src="https://user-images.githubusercontent.com/33354141/195532031-8a2dc4b3-9117-451c-b5cd-7830ce9f2ef4.png">

it would be required to breakout the hour prices somehow. I am aware you have all of that data in the attributes of (all) sensors, but that still would require a huge work on templating. Btw, those attributes are enormous, and when recorded take up a lot of space in the DB, which current HA dev's dont feel is as it should?

so, my question would be to breakout those hour prices (maybe grouped per day), so we can easily use them in hour HA systems.
I fear the best way would be to create separate sensors for each hour. (and ditch the attributes) 
A bit like the forecast sensors in lets say Buienradar... 
would only be 48 of them ;-)

Hope you will consider, thx!

**Comment by JaccoR** [comment]
If there are more people interested in having the sensor of every hour i can make an option for that. I personally don't prefer having 48 extra sensors in my home assistant system. Other than that, wouldn't having 48 sensors also take a lot of space in the DB?
It's a pity that home assistant does not have support for sensors that have future data. 

You can make a graph like that with apexcharts right? Or am i missing something in that graph?

**Reply by Mariusthvdb** [reply]
Thanks for your response!
As I understand it, it's HA policy to breakout attributes into sensors. This would follow that 'rule'

next to that, you now have all attributes on all sensors, which is way more problematic for the recovered db than those single sensors.
thats isnt necessary at all, and it would clean that up nicely.

the latter would be a FR/Discussion of mine anyways, if not touched here already ;-)

btw, there are no tomorrow sensors at all?

**Reply by ikke-zelf** [reply]
create a second graph and set the graph span to 48h.
the 2nd graph should look to: entity.attributes.prices_**tomorrow**.map

**Reply by Mariusthvdb** [reply]
right! and I was thinking we could do that in 1 graph....

wondering if we really cant though, and create that data_generator so it concatenates the 2 sets and we can have it all in one. but wait, there's more to it than that:

this is for today, and the attribute gets populated:

<img width="327" alt="Schermafbeelding 2022-10-18 om 07 32 20" src="https://user-images.githubusercontent.com/33354141/196343760-47bd6940-239f-499e-bbaa-c867f4d4cce7.png">

but look at tomorrow:

<img width="326" alt="Schermafbeelding 2022-10-18 om 07 32 11" src="https://user-images.githubusercontent.com/33354141/196343816-7f5cf0a6-eac3-4214-b49f-08713b4d8c77.png">

prices_tomorrow is empty, and there's a global prices: attribute, which is identical to the prices_today attribute ?

is that because it now is too early for the prices to be available?

also, while we're on the tomorrow subject, it would be awesome if the integration would create this:

<img width="427" alt="Schermafbeelding 2022-10-18 om 07 35 54" src="https://user-images.githubusercontent.com/33354141/196344371-c5262218-57a3-456d-95fc-2782927d5eed.png">

also for Tomorrow.... see: https://github.com/JaccoR/hass-entso-e/issues/28

**Reply by Ascotg** [reply]
@JaccoR Thanks for the great integration.
Wondering if you had any status update on the 48 sensors?

**Comment by woytekbode** [comment]
I think this would be a nice feature to have!

**Comment by Ascotg** [comment]
I found a workaround via 'Node-Red'.

When you get the object info through 'get entities' in node red you can filter on the array of prices using the following:

msg.payload[0].attributes.prices[0].price
msg.payload[0].attributes.prices[1].price
...
up to 48 hours

Once you have this you can do all sorts of things in node red:

- Sort from low to high prices
- Export as a sensor (via MQTT for instance, check out this video: https://www.youtube.com/watch?v=dqTn-Gk4Qeo)
- And more...

Here's my node red flow for this:

`[{"id":"aaf8249566ac4136","type":"inject","z":"6407955a01e5b2fa","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":260,"y":320,"wires":[["a39eda9e97a5cd11"]]},{"id":"a572c0de5d1094b4","type":"debug","z":"6407955a01e5b2fa","name":"debug 1","active":true,"tosidebar":true,"console":true,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":890,"y":320,"wires":[]},{"id":"a39eda9e97a5cd11","type":"ha-get-entities","z":"6407955a01e5b2fa","name":"","server":"9405c3fe.d0a6c","version":0,"rules":[{"property":"entity_id","logic":"starts_with","value":"sensor.epex_be_and_costs_average_electricity_price_today","valueType":"str"}],"output_type":"array","output_empty_results":false,"output_location_type":"msg","output_location":"payload","output_results_count":1,"x":470,"y":320,"wires":[["a6c18e37d3cb14aa"]]},{"id":"a6c18e37d3cb14aa","type":"function","z":"6407955a01e5b2fa","name":"function 1","func":"msg.payload = msg.payload[0].attributes.prices[0].price;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":680,"y":320,"wires":[["a572c0de5d1094b4"]]},{"id":"9405c3fe.d0a6c","type":"server","name":"Home Assistant","addon":true}]`

**Comment by uermend** [comment]
I'm looking for an automation were I can use the lowest price to switch on my waterheater. 
But my waterheater needs 5 hours a day to fully warm up so i'm want to create 5  sensors:

- lowest price hour today.
- 2nd lowest price hour today,
- 3th lowest price hour today.
-  etc 
 
I can use these sensors in an automation to switch on my waterheater for these lowest 5 hours.

does someone have a clue how to set this up with the Entso-E intergration?

Regards,

Sander

**Reply by PerpetumGW** [reply]
@uermend I would recommend to use Node-Red like the example I give above.
You can read all values like a table and write a function to rank them low to high. Use a cut-off value (e.g. 5) so at least 5 hours a day the switch is set to high.

`[{"id":"ed708ed917b9a831","type":"tab","label":"Flow 1","disabled":false,"info":"","env":[]},{"id":"bd30c2128672a919","type":"ha-get-entities","z":"ed708ed917b9a831","name":"Get energy prices today and tomorrow","server":"60f4016.b35ea","version":1,"rules":[{"property":"entity_id","logic":"is","value":"sensor.epex_be_and_costs_average_electricity_price_today","valueType":"str"}],"outputType":"array","outputEmptyResults":false,"outputLocationType":"msg","outputLocation":"payload","outputResultsCount":1,"x":350,"y":340,"wires":[["d2b01bee478acedc"]]},{"id":"d2b01bee478acedc","type":"function","z":"ed708ed917b9a831","name":"Extract array of just prices 48h","func":"// Extract prices_today array from the attributes object\nconst pricesToday = msg.payload[0].attributes.prices;\n\n// Create an array to store the price values\nconst priceValues = [];\n\n// Loop through the prices_today array and extract price values\nfor (const entry of pricesToday) {\n    priceValues.push(entry.price);\n}\n\n// Ensure the array length is always 48\n// Note when start charging before 13h00 we only have until midnight today of price values.\n// this will probably mean the EV setpoints will be likely the same: e.g. 10h00-16h00 and 21h00-23h00\nwhile (priceValues.length < 48) {\n    priceValues.push(100);\n}\n\n// Save the price values array in the message payload\nmsg.payload = priceValues;\n\n// Return the modified message\nreturn msg;\n","outputs":1,"timeout":"","noerr":0,"initialize":"","finalize":"","libs":[],"x":690,"y":340,"wires":[["d343e00a6326f230"]]},{"id":"d343e00a6326f230","type":"function","z":"ed708ed917b9a831","name":"Rank","func":"// Input array\nconst inputArray = msg.payload;\n\n// Check if the input is a valid array\nif (!Array.isArray(inputArray)) {\n    node.error(\"The input should be a valid array.\", msg);\n    return null;\n}\n\n// Sort the array in ascending order\nconst sortedArray = inputArray.slice().sort((a, b) => a - b);\n\n// Generate the ranks array\nconst ranksArray = inputArray.map((value) => sortedArray.indexOf(value) + 1);\n\n// Save the ranks array in the message payload\nmsg.payload = ranksArray;\n\n// Return the modified message\nreturn msg;\n","outputs":1,"timeout":"","noerr":0,"initialize":"","finalize":"","libs":[],"x":930,"y":340,"wires":[["03f5a1fa3b4018af"]]},{"id":"03f5a1fa3b4018af","type":"function","z":"ed708ed917b9a831","name":"Generate array of the 8 most suitable times","func":"// Define the threshold (replace this with your desired value)\nconst threshold = 9;\n\n// Function to set rows with values higher than the threshold to zero\nfunction setRowsToZero(arr) {\n    const result = arr.map((value) => {\n        // If the current value is higher than the threshold, set it to zero\n        if (value > threshold) {\n            return 0;\n        } else {\n            return 1;\n        }\n    });\n    return result;\n}\n\n// Check if the payload is an array\nif (Array.isArray(msg.payload)) {\n    // Call the setRowsToZero function and update the payload\n    msg.payload = setRowsToZero(msg.payload);\n} else {\n    // Handle the case when the payload is not an array (optional)\n    msg.payload = [];\n}\n\ncontext.flow.set(\"thegoodtimes\", msg.payload);\n\n// Return the modified message\nreturn msg;\n","outputs":1,"timeout":"","noerr":0,"initialize":"","finalize":"","libs":[],"x":1210,"y":340,"wires":[["7d234f99efde0368","5764813d0889a184"]]},{"id":"7d234f99efde0368","type":"debug","z":"ed708ed917b9a831","name":"debug 1","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":1500,"y":300,"wires":[]},{"id":"25034e0ca7c3ec8c","type":"inject","z":"ed708ed917b9a831","name":"Poll every 20 sec","props":[{"p":"payload"}],"repeat":"20","crontab":"","once":true,"onceDelay":"1","topic":"","payload":"","payloadType":"date","x":150,"y":240,"wires":[["bd30c2128672a919"]]},{"id":"5764813d0889a184","type":"function","z":"ed708ed917b9a831","name":"Array of timeslots","func":"// Initialize an array to store timestamps\nconst start = new Date(); // Today at midnight\nstart.setHours(0, 0, 0, 0);\n\n// Helper function to format date to local ISO-like string\nfunction toLocalTimestamp(date) {\n    const year = date.getFullYear();\n    const month = String(date.getMonth() + 1).padStart(2, '0'); // Months are 0-based\n    const day = String(date.getDate()).padStart(2, '0');\n    const hours = String(date.getHours()).padStart(2, '0');\n    const minutes = String(date.getMinutes()).padStart(2, '0');\n    const seconds = String(date.getSeconds()).padStart(2, '0');\n    const milliseconds = String(date.getMilliseconds()).padStart(3, '0');\n\n    return `${year}-${month}-${day}T${hours}:${minutes}:${seconds}.${milliseconds}`;\n}\n\n// Create an array of timestamps\nconst timestamps = [];\nfor (let i = 0; i < 48; i++) {\n    const next = new Date(start.getTime() + i * 3600000); // Add hours\n    timestamps.push(toLocalTimestamp(next)); // Convert to local ISO-like string\n}\n\n// Pair timestamps with array values\nconst results = timestamps.map((time, index) => ({\n    time: time,\n    value: msg.payload[index]\n}));\n\n// Filter the results to only include timestamps where value is 1\nconst filteredTimestamps = results\n    .filter(entry => entry.value === 1) // Keep only entries where value is 1\n    .map(entry => entry.time); // Extract the timestamp\n\n// Return the filtered timestamps\nmsg.payload = filteredTimestamps;\nreturn msg;\n","outputs":1,"timeout":0,"noerr":0,"initialize":"","finalize":"","libs":[],"x":1510,"y":400,"wires":[["9c61de44fc171ebf"]]},{"id":"9c61de44fc171ebf","type":"debug","z":"ed708ed917b9a831","name":"debug 2","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":1660,"y":500,"wires":[]},{"id":"60f4016.b35ea","type":"server","name":"Home Assistant","addon":true,"rejectUnauthorizedCerts":true,"ha_boolean":"","connectionDelay":false,"cacheJson":false,"heartbeat":false,"heartbeatInterval":"","statusSeparator":"","enableGlobalContextStore":false}]`
